GVRf/Framework/backend_oculus/src/main/java/org/gearvrf/OvrGearController.java

```diff
@@ -0,0 +1,352 @@
+import android.os.HandlerThread;
+import android.os.Message;
+import android.view.KeyEvent;
+
+import org.gearvrf.io.CursorControllerListener;
+import org.gearvrf.io.GVRControllerType;
+
+import org.gearvrf.io.GVRInputManager;
+import org.joml.Quaternionf;
+import org.joml.Vector3f;
+
+import java.nio.ByteBuffer;
+import java.nio.ByteOrder;
+import java.nio.FloatBuffer;
+
+/**
+ * This class represents the Gear Controller.
+ *
+ * The input manager notifies the application when the controller has connected successfully. Use
+ * the {@link GVRInputManager#addCursorControllerListener(CursorControllerListener)} to get notified
+ * when the controller is available to use. To query the device specific information from the
+ * Gear Controller make sure to type cast the returned {@link GVRCursorController} to
+ * {@link OvrGearController} like below:
+ *
+ * <code>
+ * OvrGearController controller = (OvrGearController) gvrCursorController;
+ * </code>
+ *
+ * Additionally register a {@link ControllerEventListener} using
+ * {@link GVRCursorController#addControllerEventListener(ControllerEventListener)} to receive
+ * notification whenever the controller information is updated.
+ */
+public class OvrGearController extends GVRCursorController implements GVRDrawFrameListener {
+    private static final String TAG = OvrGearController.class.getSimpleName();
+
+    private static final int DATA_SIZE = 12;
+    private static final int BYTE_TO_FLOAT = 4;
+    private static final int INDEX_CONNECTED = 0;
+    private static final int INDEX_HANDEDNESS = 1;
+    private static final int INDEX_POSITION = 2;
+    private static final int INDEX_ROTATION = 5;
+    private static final int INDEX_BUTTON = 9;
+    private static final int INDEX_TOUCHPAD = 10;
+    private static final int OVR_BUTTON_A = 0x00000001;
+    private static final int OVR_BUTTON_ENTER = 0x00100000;
+    private static final int OVR_BUTTON_BACK = 0x00200000;
+
+    /**
+     * Defines the handedness of the gear controller.
+     */
+    public enum Handedness {
+        LEFT, RIGHT
+    }
+
+    private GVRSceneObject pivot;
+    private GVRContext context;
+    private FloatBuffer readbackBuffer;
+
+    private EventHandlerThread thread;
+    private boolean initialized = false;
+    private final long mPtr;
+
+    OvrGearController(GVRContext context) {
+        super(GVRControllerType.CONTROLLER);
+        ByteBuffer readbackBufferB = ByteBuffer.allocateDirect(DATA_SIZE * BYTE_TO_FLOAT);
+        readbackBufferB.order(ByteOrder.nativeOrder());
+        readbackBuffer = readbackBufferB.asFloatBuffer();
+        mPtr = OvrNativeGearController.ctor(readbackBufferB);
+        this.context = context;
+        pivot = new GVRSceneObject(context);
+        thread = new EventHandlerThread();
+        context.registerDrawFrameListener(this);
+    }
+
+    long getPtr() {
+        return mPtr;
+    }
+
+    @Override
+    public void setSceneObject(GVRSceneObject object) {
```

Comment:
The base class exposes resetSceneObject method that calls setSceneObject with null parameter. Would calling resetSO work as expected for OvrGearController?